algebra/monoid.lean

```diff
@@ -0,0 +1,77 @@
+
+import category.basic
+import category_theory.category category_theory.types
+open category_theory
+
+universes u v
+
+def dual (α : Sort u) := α
+
+def dual.mk {α : Sort u} (x : α) : dual α := x
+
+def Monad (m) [monad.{u v} m] := Type u
+
+def Monad.mk (m) [monad.{u v} m] (α : Type u) : Monad m := α
+
+instance {m} [monad m] : has_hom (Monad m) :=
+{ hom := λ α β, α → m β,
+  id := λ α x, (pure x : m α),
+  comp := λ X Y Z f g, f >=> g }
+
+instance {m} [monad m] [is_lawful_monad m] : category (Monad m) :=
+by refine { hom := λ α β, α → m β,
+            id := λ α x, (pure x : m α),
+            comp := λ X Y Z f g, f >=> g,
+            id_comp' := _, comp_id' := _, assoc' := _ };
+   intros; ext; simp only [(>=>)] with functor_norm
+
+@[simp] lemma endo.one_def {α : Type u} : (1 : End α) = @id α := rfl
+
+@[simp] lemma endo.mul_def {α : Type u} (xs ys : End α) : xs * ys = ys ∘ xs := rfl
+
+@[simp] lemma monad.one_def {m} [monad m] [is_lawful_monad m] (α : Monad m) :
+  (1 : End α) = @pure m _ α := rfl
+
+@[simp] lemma monad.mul_def {m} [monad m] [is_lawful_monad m] (α : Monad m) (xs ys : End α) :
+  xs * ys = xs >=> ys := rfl
+
+instance {α} [has_zero α] : has_zero (dual α) :=
+{ zero := (0 : α) }
+
+instance {α} [has_add α] : has_add (dual α) :=
+{ add  := λ x y, (y + x : α) }
+
+instance {α} [add_monoid α] : add_monoid (dual α) :=
+{ zero := (0 : α),
+  add  := λ x y, (y + x : α),
+  add_zero := @zero_add α _,
+  zero_add := @add_zero α _,
+  add_assoc := by intros;symmetry;apply @add_assoc α }
+
+instance {α} [has_one α] : has_one (dual α) :=
```

Comment:
Are you sure we want to call this `dual`? `opposite` would be consistent with category theory.

Reply:
Nice! Thanks! Can I use opposite on anything I like or is it mostly used on morphisms or functors or other?

Reply:
[`opposite`](https://github.com/leanprover/mathlib/blob/master/category_theory/opposites.lean#L12) has the same definition as `dual`, although on `Type u`, rather than `Sort u`. (Perhaps that can be generalised.)

It's intended for use on anything --- opposite groups, opposite manifolds, etc.